Recently Edited: Merge duplicate fix

shresthabijay opened the pull request:
refers to #323

shresthabijay commented:
Hi. I have made these changes to the data structure and merge rules. For now I have used the object structure instead of tree. With this implementation we only have to iterate the list once for merging and deleting O(n). It removes all the duplicates.

Yet to implement replacing descendants if the descendant has not been edited in a long time. We need a proper way to removing direct sibling relations between new context when we replace descendants. 

Please review this approach and let me know if all the merging conditions are correctly working. I still am working on tree structure on sidelines.

Ogutu-Brian commented on src/reducers/existingThoughtChange.js:
```diff
@@ -66,37 +65,58 @@ export default (state, { oldValue, newValue, context, showContexts, thoughtsRank
-        // this variable sets to true when there is atleast one recently edited thought it can merge to
-        if (subcontextIndex > -1 && timeDiffInSec <= 7200) isMerged = true // eslint-disable-line fp/no-mutating-methods
+  if (updatedRecentlyEdited[oldPathIndex]) {
+    delete updatedRecentlyEdited[oldPathIndex] // eslint-disable-line fp/no-delete
```
Please do not use delete in arrays, it will leave holes that might be so hard to debug. 
`updatedRecentlyEdited[oldPathIndex]` would have a value of undefined which I really don't think is good. 
`updatedRecentlyEdited = updatedRecentlyEdited.filter((updatedValue, index)=>index !== oldPathIndex)` would give you a refined array instead

Ogutu-Brian replied:
https://rules.sonarsource.com/javascript/RSPEC-2870

raineorshine replied:
@Ogutu-Brian `updatedRecentlyEdited` is an object, not an array.

Ogutu-Brian replied:
Thanks @raineorshine ,
I didn't know that.

raineorshine commented:
It seems that this will be substantially different with the tree implementation, so maybe I will wait until then to review, given the complexity of this code. Do you still need me to review before implementing the tree structure?

shresthabijay commented:
> It seems that this will be substantially different with the tree implementation, so maybe I will wait until then to review, given the complexity of this code. Do you still need me to review before implementing the tree structure?

No it won't be necessary to review this right now . I will complete the tree implementation first.

shresthabijay commented:
Hey Raine. I have completed the tree structure implementation. I have written logic for almost all of the merge cases we discussed earlier. I have tested on my own . There are no duplicates anymore and merges are happening properly. I have written logic for replacing descendant context with the ancestor if descendant thoughts are too old too. I still have to work on limiting no of recently edited thoughts in the tree. Please test the merge cases, deletion and ancestor replacing descendant cases thoroughly and let know if everything is happening as we discussed. Also let me know about the possible changes you want me to make to the codebase.

raineorshine commented:
Haven't looked at the code yet, but functionally this is working great! Everything is merging very intuitively. I've tested some of our original use cases, plus some random adding, editing, and deleting. Very impressive!

I'd like to try this "in the wild" and provide more feedback on the behavior after I have used it with real data. This is already a big improvement though.

The only thing that seemed off was moving:

e.g.

- a
  - b
  - c
    - d
      - e

The recently edited list is `a.c.d.e`. When I move `c` before `b`, it becomes `a.c`. I would have expected it to still be `a.c.d.e`.

Similar behavior occurs when dragging into a different context, but the expected behavior is not as obvious, so I am not sure yet if I consider that an issue.

raineorshine commented on src/util/initialState.js:
```diff
@@ -42,7 +42,7 @@ export const initialState = () => {
       },
     },
-    recentlyEdited: [],
+    recentlyEdited: { _ROOT_: {} },
```
Is this supposed to be the same as `ROOT_TOKEN`, or is this independent? Either way, please use a constant.

shresthabijay replied:
Sure

raineorshine commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,115 @@
+  return contextIndex > -1 ? { node: _.get(tree, context.slice(0, context.length - contextIndex + 1)), path: context.slice(0, context.length - contextIndex + 1) } : {}
+}
+
+export const onNodeChange = (tree, oldPath, newPath) => {
```
Please return a new tree instance instead of modifying the argument reference.

shresthabijay replied:
Sure. I hope cloning the original tree object and modifying it instead would do it.

raineorshine replied:
No, a deep clone is unnecessary. It'll be cleaner to simply add or replace the appropriate subtree.

raineorshine commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,115 @@
+
+const editTimeMax = 30 // time diff limit in second for replacing descendants by ancestor
+
+const findTreeDeepestSubcontext = (tree, path) => {
```
I recommend adding a jsdoc style comment above function with a description of what it does. Even with descriptive names, this is a nice convention that will help other developers and allow for auto-generated docs in the future.

shresthabijay replied:
Sure.

raineorshine commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,115 @@
+  const node = _.get(tree, startingPath)
+  if (!node) return []
+  if (node.leaf) leafNodes.push({ ...node }) // eslint-disable-line fp/no-mutating-methods
+  else Object.keys(node).forEach(child => findTreeDescendants(tree, [...startingPath, child], leafNodes))
```
Let's use `reduce` here instead of mutation.

shresthabijay replied:
Yes it makes sense.

raineorshine commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,115 @@
+    const node = _.get(tree, (context.slice(0, context.length - index)))
+    return !node.leaf && Object.keys(node).length > 1
+  })
+  return contextIndex > -1 ? { node: _.get(tree, context.slice(0, context.length - contextIndex + 1)), path: context.slice(0, context.length - contextIndex + 1) } : {}
```
I recommend placing each `key: value` pair on a separate line to help readability with all but the smallest object declarations.

raineorshine commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,115 @@
+  }
+}
+
+export const onNodeDelete = (tree, oldPath) => {
```
Let's call these methods `treeDelete` and `treeChange`. `onSomething` indicates an event handler, which is not what this is.

raineorshine commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,115 @@
+  const availableNode = _.get(tree, path)
+  if (availableNode) return { node: availableNode, path }
+  const pathIndex = path.findIndex((value, index) => _.get(tree, (path.slice(0, path.length - index))))
+  return pathIndex > -1 ? { node: _.get(tree, path.slice(0, path.length - pathIndex)), path: path.slice(0, path.length - pathIndex) } : {}
```
This approach has a lot of overhead due to calling `_.get` in every iteration of `path.findIndex`. Instead, I would suggest traversing the `tree` manually, stopping as soon as a non-matching subpath is encountered, and thus returning the longest matching subpath.

raineorshine commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,115 @@
+  const contextIndex = context.findIndex((value, index) => {
+    const node = _.get(tree, (context.slice(0, context.length - index)))
+    return !node.leaf && Object.keys(node).length > 1
+  })
```
Similar to `findTreeDeepestSubcontext`, this approach has a lot of overhead due to calling `_.get` in every iteration of `context.findIndex`. Instead, traverse manually. 

`_.get` is great for one-off deep object access, but not as much for multiple adjacent lookups, i.e.:

```
_.get(tree, 'a.b.c.d.e')
_.get(tree, 'a.b.c.d')
_.get(tree, 'a.b.c')
_.get(tree, 'a.b')
_.get(tree, 'a')
```

In this case `_.get` is doing the same work over and over again unnecessarily.

shresthabijay replied:
Yes I agree. We should traverse manually.

raineorshine commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,115 @@
+  if (commonNode) {
+    if (oldContext.length > commonPath.length) return
+    if (commonPath.length !== 1) _.unset(tree, commonPath) // preventing deleting of _ROOT_
+    const parentPath = commonPath.slice(0, commonPath.length - 1)
```
Let's use the existing `contextOf` method to get the `parentPath`. That offers some decoupling between the context functionality and implementation. Admittedly there are many places in the code that "know" that paths are arrays, and often need to, but it's still good to stick to the `contextOf` and `head` interfaces when appropriate to minimize this implementation-reliance.

shresthabijay replied:
Okay

raineorshine commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,115 @@
+    if (oldContext.length > commonPath.length) return
+    if (commonPath.length !== 1) _.unset(tree, commonPath) // preventing deleting of _ROOT_
+    const parentPath = commonPath.slice(0, commonPath.length - 1)
+    const grandParentPath = commonPath.slice(0, commonPath.length - 2)
```
Explicitly referencing grandparents is a bit of a red flag. What about parents of grandparents? This seems to suggest some recursive functionality is needed. It's possible I am not understanding what is happening here however, as I am still figuring out the code.

shresthabijay replied:
Right now code is bit complex and it may take a while to figure out some logic just by analyzing the code. I will write down a document explaining `treeChange` and `treeDelete` in depth. I am sure it will give you a better understanding of my perspective. Then we can continue proper discussion about further changes required.

raineorshine replied:
Okay, thanks

raineorshine commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,115 @@
+    if (commonNode.leaf) {
+      // A is changed to AF --> AF , A and A.B --> A.B , A.B and A.B.C --> A.B.C
+      _.unset(tree, commonPath)
+      _.set(tree, newContext, { leaf: true, lastUpdated: timestamp(), path: newPath })
```
Let's write our own non-destructive versions of `_.set` and `_.unset`. I'm really not comfortable with these mutations.

Note: They will need to be non-destructive at all depths, so a shallow copy won't work (and a deep copy is unnecessary).

shresthabijay replied:
Currently I am modifying the same tree object. Are you suggesting that we manually traverse tree up to the required path to add or delete index ?

raineorshine replied:
Yes

shresthabijay commented:
Hey Raine. In this latest commit I have changed the `class` implementation back to functional and used `Immer` for immutability. Immer `produce` allows mutatation to its `draftState` and at end it will produce the `nextState`  based on the mutations to the draft. We already completed writing logic for `treeChange`, `treeDelete` that mutates the tree . So we just wrap them with Immers `produce`  and provide `draftState` to them instead of original tree and even after all the mutations we will get pure immutable tree object in return from `produce`. **Works like magic.** 

**I have completed move logic too but it still needs to be tested properly.** 

All the requested changes are not in this latest commit though. I will work on those changes after you give me green flag on Immer implementation and move logic test.

raineorshine commented on src/components/Sidebar.js:
```diff
@@ -20,7 +23,7 @@ const ThoughtsTab = ({ thoughtsRanked }) => {
 
 const RecentEdited = () => {
-  const recentlyEdited = _.sortedUniqBy(_.reverse(_.sortBy(useSelector(state => (state.recentlyEdited)), 'lastUpdated')), recentThought => hashContext(recentThought.path)) // eslint-disable-line fp/no-mutating-methods
+  const recentlyEdited = _.sortedUniqBy(_.reverse(_.sortBy(findTreeDescendants(useSelector(state => (state.recentlyEdited)), [EM_TOKEN]), 'lastUpdated')), recentThought => hashContext(recentThought.path)) // eslint-disable-line fp/no-mutating-methods
```
We can remove `_.sortedUniqBy` since the tree eliminates duplicates now.

Should also be able to get rid of the `eslint-disable-line` as well.

shresthabijay replied:
Sure.

raineorshine commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,153 @@
+
+export const treeDelete = (originalTree, oldPath) => produce(originalTree, draftTree => nodeDelete(draftTree, oldPath))
+
+export const treeMove = (originalTree, oldPath, newPath) => produce(originalTree, draftTree => nodeMove(draftTree, oldPath, newPath))
```
Using some FP magic, we can abstract the pattern of using immer on the first argument:

```
/** Use immer to pass a draft object as the first argument to the given destructive function to avoid mutation **/
const immerfy = f => (obj, ...rest) => produce(obj, draft => f(draft, ...rest))

export const treeChange = immerfy(nodeChange)
export const treeDelete = immerfy(nodeDelete)
export const treeMove = immerfy(nodeMove)
```

shresthabijay replied:
Agreed!

raineorshine commented on src/reducers/existingThoughtChange.js:
```diff
@@ -69,38 +64,7 @@ export default (state, { oldValue, newValue, context, showContexts, thoughtsRank
-      })
-      .concat(isMerged ? [] : [{ path: newPath, lastUpdated: timestamp() }]),
-    'lastUpdated'))
+  const recentlyEdited = treeChange({ ...state.recentlyEdited }, oldPath, newPath)
```
We should be able to pass `state.recentlyEdited` directly now since immer makes it non-destructive, yes?

shresthabijay replied:
Yes.

raineorshine commented on src/util/initialState.js:
```diff
@@ -42,7 +42,7 @@ export const initialState = () => {
       },
     },
-    recentlyEdited: [],
+    recentlyEdited: { [EM_TOKEN]: {} },
```
I think the `EM_TOKEN` is extraneous in the tree. It would be cleaner to initialize to `{}` and then `_.set` directly on the object rather than always in `EM_TOKEN`.

shresthabijay replied:
Agreed!

raineorshine commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,153 @@
+  const node = tree[path[index]]
+  if (node) return findTreeDeepestSubcontext(node, path, index + 1)
+  else return { node: index === 0 ? null : tree, path: path.slice(0, index) }
+}
```
Nice improvement!

shresthabijay replied:
Thanks!

raineorshine commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,153 @@
+  if (!node) return []
+  else if (node.leaf) return [{ ...node }]
+  else return Object.keys(node).reduce((acc, child) => [...acc, ...findTreeDescendants(tree, [...startingPath, child])], [])
+}
```
This still needs some work. Notice that you are still calling `_.get` each iteration. This results in complexity of `O(depth) * O(depth)`. `_.get` is great when a single node is accessed, but it shouldn't be used in a loop or iteration, since it will repeat work. Instead of you want a `O(n)` recursive traversal function. Let me know if you need more guidance on this.

shresthabijay replied:
Yes understood. Done!

shresthabijay commented:
In this latest change I have fixed _.get iteration issue, added jsdocs style comments and other requested changes. I also completed move logic. I have tested and merges are happening intuitively. I want you review the functionality properly specially **deleting** and **moving**. Let me know soon!

raineorshine commented:
Minor change, but could you add a shortcut for toggling the sidebar?

`toggleSidebar.js:`

```
import { store } from '../store.js'

export default {
  id: 'toggleSidebar',
  name: 'Toggle Recently Edited',
  keyboard: { alt: true, key: 'r' },
  hideFromInstructions: true,
  exec: () => {
    store.dispatch({ type: 'toggleSidebar', value: !store.getState().showSidebar })
  }
}
```

raineorshine commented:
This is working so well! I only encountered one edge case and one crash that sadly I cannot reproduce.

The edge case involves the `subcategorizeAll`. Start with these thoughts:

- a

shresthabijay commented:
Hi I have fixed the **subcategorizeAll** issue and also refactored some functions in this latest changes. Please review the changes and functionality. Thank you!

raineorshine commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,214 @@
+          const longContext = descendantContext.length > newContext.length ? descendantContext : newContext
+          const shortContext = newContext.length < descendantContext.length ? newContext : descendantContext
+          // siblings A.B.C and A.B.D ---> A.B.D
+          if (descendantContext.length === newContext.length && newContext.length - commonPath.length === 1) {
```
You may want to consider creating constants for parts of conditionals to make them more readable. (Tell me if the names don't make sense here, I could be misunderstanding them, but you get the point.)

```js
const isSameDepth = descendantContext.length === newContext.length
const hasSameParent = newContext.length - commonPath.length === 1
const isSibling === isSameDepth && hasSameParent
...
```

This can be applied to other places to make the different conditions a bit clearer.

shresthabijay replied:
It makes perfect sense. I will make these changes to other places.

shresthabijay commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,214 @@
+  if (context.length === 0 && index === 0) return tree
+  const node = tree[context[index]]
+  if (node) return findTreeDeepestSubcontext(node, context, index + 1)
+  else return { node: tree, path: context.slice(0, index) }
```
Agreed!

shresthabijay commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,214 @@
+   * @returns {string[]} common subcontext
+   */
+const findTreeDeepestSubcontext = (tree, context, index = 0) => {
+  if (context.length === 0 && index === 0) return tree
```
Sure!

shresthabijay commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,214 @@
+  if (commonNode) {
+    if (oldContext.length > commonPath.length) return // if node is not already available in the tree
+    const closestAncestor = findClosestAncestorWithMultipleChildren(tree, commonPath)
+    const nodeToMergeIntoPath = commonPath.length - closestAncestor.path.length <= 2 ? closestAncestor.path : false
```
Agreed!

shresthabijay commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,214 @@
+  if (!node) return []
+  else if (node.leaf) return [{ ...node }]
+  else return Object.keys(node).reduce((acc, child) => [...acc, ...findTreeDescendants(node, startingPath, child)], [])
+}
```
That's a better implementation !

shresthabijay commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,214 @@
+   * @param {string[]} context array of string representing path
+   * @returns {Object} closest ancestor node with multiple children
+   */
+const findClosestAncestorWithMultipleChildren = (tree, context, minChildren = 2, index = 0, closestAncestor = { node: null, path: [] }) => {
```
Sure! It makes perfect sense to call it `closestSharedAncestor` .

shresthabijay commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,214 @@
+  const node = tree[context[index]]
+  if (node) return findClosestAncestorWithMultipleChildren(node, context, minChildren, index + 1, (!node.leaf && Object.keys(node).length >= minChildren) ? { node, path: context.slice(0, index + 1) } : closestAncestor)
+  else return closestAncestor
+}
```
Agreed!

shresthabijay commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,214 @@
+   * @param {string[]} context array of string representing path
+   * @returns {string[]} common subcontext
+   */
+const findTreeDeepestSubcontext = (tree, context, index = 0) => {
```
It will return root tree itself . It is because even when new context that has no ancestor at all in the tree at is given it could be a possible new leaf node. For example:

-A
-B

For example if we have these thoughts in the main tree, recently edited list will be

-B  (Since A and B have direct siblings relation , A and B will be merged to be just B)

Now when C is added as `subthought` of A then A.C will have no common path in the recently edited tree at all. But it is a possible case of new thought being added to the recently edited tree (A.C and B will must merge to become A.C ). Therefore we need return the main tree as the `commonNode`.

shresthabijay commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,214 @@
+      })
+    }
+    // if cannot be merged to either parent or grandParent then just making parent path as the leaf node *** exception if parent path is __ROOT__ node ***
+    else {
```
Sure

shresthabijay commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,214 @@
+      const pathBeingMerged = commonPath.slice(0, closestAncestor.path.length + 1)
+      _.unset(tree, pathBeingMerged) // deleting the merged path
+      findTreeDescendants(tree, nodeToMergeIntoPath).forEach(descendant => {
+        _.set(tree, pathToContext(descendant.path), { leaf: true, lastUpdated: timestampUpdate ? timestamp() : descendant.lastUpdated, path: descendant.path })
```
I updated it before because `A.B.C.D.E` is the changed path and it has now distant relation with these `A.B.C.G.H.K` and `A.B.C.G.H.I.J` . So it's actually merging to these paths. But I too think we should not update the timestamp in case of delete. Should we change it?

raineorshine replied:
I've thought about it some more, and let's leave it for now. There is some reason to it, since   `A.B.C.G.H.I.J` and `A.B.C.G.H.K` are the closest ancestors, and if they are not updated then `A.B.C.*` will not be accessible from the top of the list. I can assess more after actual usage. Thanks!

shresthabijay commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,214 @@
+    }
+    else {
+      if (oldNode.leaf) {
+        nodeChange(tree, contextOf(newPath).concat({ value: '' }), newPath)
```
Sure. If `oldContext` is already available node on the recently edited tree `if(equalArrays(commonPath, oldContext))` then `nodeChange` will only update it's descendants. When we move nodes in the tree we delete `oldNode` and add it to destination branch. So we add a empty node to the `contextOf` new path to add it as new node to the new branch . 

For example:

A.B.C.D.E.F
A.B.C.G.H.I

If we want to add A.B.C.K.L , then ['A','B','C',''] will be `oldPath` and  ['A','B','C','K','L'] will be `newPath`.

raineorshine replied:
Okay, thanks for explaining! Can we add an explicit `nodeAdd` method? It's not ideal to have a magic value like `{ value: '' }` that is used to activate a particular code path. I see too realistic options:

1. Hide the magic value in a `nodeAdd` method that delegates to `nodeChange` using `{ value: DUMMY_TOKEN }` to make it explicit. Add a comment to explain how the unconventional use of `nodeChange` achieves the desired effect.
2. Write separate `nodeAdd` and `nodeChange` methods, and abstract the common logic (may involve unnecessary effort at this point.)

shresthabijay replied:
The problem with this approach is that we are using `existingThoughtChange` and when new empty thought is added and first value is entered we get same pattern of `oldPath` and `newPath` by `existingThoughtChange`.

For example:
If new empty `subthought` is added in A.B.C.D and we write value E inside the new empty thought. We get `oldPath` ['A','B','C','D',''] and `newPath` ['A','B','C','D','E']. 

What I mean to say is that we don't know if a new thought has been added to the tree because recently edited thoughts methods are called only on `existingThoughtChange`. Also we add nodes to the recently edited tree only after `nodeChange` doesn't find already available nodes. If it finds one it updates it's descendants **(change)** and if node is not available **(possible new node)** it makes sure that there are no merge able relations between already available nodes before adding a new leaf to the branch. So due to the fact that we don't know if new path will be added to the recently edited tree as leaf node, we have to first check for `nodeChange` and only then we can perform addition of new leaf node to the tree.

In case of node move we have to create a `oldPath` and a `newPath` for moved branches. Since `nodeChange` works as I explained above example. We have to provide same pattern of `oldPath` and `newPath` . So we have to add `{value:''}`. 

I think we should go with your first suggestion and write a function `nodeAdd` that adds `{ value: DUMMY_TOKEN }` to the provided `oldPath` and explain the logic there. `nodeAdd` will be used where we have to imitate new thought being add.

shresthabijay commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,214 @@
+          }
+        })
+        if (descendantReplaced) nodeChange(tree, oldPath, newPath) // called once again to remove merge inconsitencty that might occur while replacing descendants by ancestor
+        else if (oldContext.slice(-1)[0] !== newContext.slice(-1)[0]) _.unset(tree, oldContext)
```
Sure

shresthabijay commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,214 @@
+      if (equalArrays(commonPath, oldContext)) {
+        // this flag is needed to know if ancestor replaces very old descentdant in the tree
+        let descendantReplaced = false // eslint-disable-line fp/no-let
+        // If A.E which already available in tree is updated to A.EM the updating just its descendants like A.E.O.M to A.EM.O.M.
```
Sure. It's for directly updating descendants for already available node in the recently edited tree.

`
Recently Edited List

          A.B.C.D.E.F
          A.Q.K.R
          A.B.G.H.I

          if A.B is changes to A.BE then directly updating all its desendants

          A.BE.C.D.E.F
          A.BE.G.H.I
          A.Q.K.R
`

shresthabijay commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,214 @@
+        let isMerged = false // eslint-disable-line fp/no-let
+        leafNodes.forEach(descendant => {
+          const descendantContext = pathToContext(descendant.path)
+          if (descendantContext[0] === '__EM__') return // preventing nodes at level 0 from merged to this (temporary fix)
```
understood

shresthabijay commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,214 @@
+          const descendantContext = pathToContext(descendant.path)
+          if (descendantContext[0] === '__EM__') return // preventing nodes at level 0 from merged to this (temporary fix)
+          const longContext = descendantContext.length > newContext.length ? descendantContext : newContext
+          const shortContext = newContext.length < descendantContext.length ? newContext : descendantContext
```
Sure ! will do that

shresthabijay commented on src/components/Sidebar.js:
```diff
@@ -20,7 +16,7 @@ const ThoughtsTab = ({ thoughtsRanked }) => {
 
 const RecentEdited = () => {
-  const recentlyEdited = _.sortedUniqBy(_.reverse(_.sortBy(useSelector(state => (state.recentlyEdited)), 'lastUpdated')), recentThought => hashContext(recentThought.path)) // eslint-disable-line fp/no-mutating-methods
+  const recentlyEdited = _.reverse(_.sortBy(findTreeDescendants(useSelector(state => (state.recentlyEdited)), []), 'lastUpdated')) // eslint-disable-line fp/no-mutating-methods
```
`_.reverse` is treated as a mutating method by eslint . So we need it to disable the error.

raineorshine replied:
Interesting! I had assumed it was non-destructive.

raineorshine commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,286 @@
+  const { node: commonNode, path: commonPath } = findTreeDeepestSubcontext(tree, oldContext)
+  if (commonNode) {
+    const isNodeNotAvailable = oldContext.length > commonPath.length
+    if (isNodeNotAvailable) return // if node is not already available in the tree
```
Here I would simply inline the conditional since it is unitary. Intermediate variables are justified only when you have a complex composite conditional, such as `if (A && (B || C))` or in some cases just `if (A && B)` if the expressions are long enough.

shresthabijay replied:
Agreed!

raineorshine commented:
> Minor change, but could you add a shortcut for toggling the sidebar?
> 
> `toggleSidebar.js:`
> 
> ```
> import { store } from '../store.js'
> 
> export default {
>   id: 'toggleSidebar',
>   name: 'Toggle Recently Edited',
>   keyboard: { alt: true, key: 'r' },
>   hideFromInstructions: true,
>   exec: () => {
>     store.dispatch({ type: 'toggleSidebar', value: !store.getState().showSidebar })
>   }
> }
> ```

@shresthabijay

shresthabijay commented:
Hi! I have added `nodeAdd` function that just calls `nodeChange` but is more structured for special case of adding new node. I have added important comments and examples to clarify how it is working. I also added created a `DUMMY_TOKEN` instead of direct use of empty string in `nodeAdd`.

I also added `toggleSidebar` `shortcut`. But `alt+r` was not working somehow , so i used `alt+e` for now. 

Please review the changes. Thank you!

shresthabijay commented:
I totally forgot about this change. It's done in the latest commit !

raineorshine commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,307 @@
+      nodeChange(tree, oldPath, newPath)
+    }
+    else if (oldNode.leaf) {
+      nodeAdd(tree, contextOf(newPath), [head(newPath)])
```
`nodeAdd(tree, newPath)`

raineorshine commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,307 @@
+      const descendants = findTreeDescendants(tree, oldContext)
+      descendants.forEach(descendant => {
+        const addedPath = [head(newPath)].concat(descendant.path.slice(oldPath.length))
+        nodeAdd(tree, contextOf(newPath), addedPath)
```
`nodeAdd(newPath.concat(descendant.path.slice(oldPath.length))`

*This differs slightly from the current implementation. I think it will still work but correct me if I am wrong. I am hoping that `nodeAdd` can be used the same way in each conditional branch, albeit with different arguments, so that it does not have to have a special flag for this case.

shresthabijay replied:
About this I just realized that with the current implementation of `nodeChange` , we can add new node by passing `newPath` as both `oldPath` and `newPath` parameter of `nodeChange`. It is because `commonPath` will be be derived from `oldPath` anyways. So this change will eliminate the need of `DUMMY_TOKEN` too.

raineorshine replied:
Aha! My suspicion was correct :). I had assumed `oldPath` was derivable from `newPath`, but I didn't know it would be as easy as passing `newPath` as `oldPath` given the way that `commonPath` is found. Looks great.

raineorshine commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,307 @@
+      nodeDelete(tree, oldPath, false)
+    }
+  }
+  else nodeAdd(tree, contextOf(newPath), [head(newPath)]) // if exact node is not found in the tree
```
`nodeAdd(tree, newPath)`

raineorshine commented on src/util/recentlyEditedTree.js:
```diff
@@ -0,0 +1,307 @@
+   * @param {string[]} addedPath array of string representing path
+   *
+*/
+const nodeAdd = (tree, oldPath, addedPath) => nodeChange(tree, contextOf(oldPath).concat({ value: DUMMY_TOKEN }), oldPath.concat(addedPath))
```
Here's another lesson about encapsulation I encourage you to think about.

When designing the signature for a function, ask yourself the questions:

- What is the minimum the caller needs to know to call this function?
- What is a natural expression of the inputs of this function?

Intuitively, the inputs for adding a node should simply be the tree and the node to add. The signature of `(tree, oldPath, addedPath) => ...` expects the caller to know something about `oldPath` and `addedPath` which is odd because it is related to the internal implementation of `nodeAdd` and not to the semantics of the caller.

Let me know if this makes sense!

shresthabijay replied:
Hi Raine. It makes perfect sense. The problem here is `nodeAdd` is just a special case of `nodeChange` . So it needs `oldPath` that is the already available concat with `{value:""}` in the tree and `newPath` that is concat of `oldPath` and new branch added to that node.

Example 

  A.B.C.D.E.F
  A.B.C.G.H.I

  If we want to add `A.B.C.K.L` , then `['A','B','C','']` will be `oldPath` and `['A','B','C','K','L']` will be 
  `newPath`.

  So node add takes ['A','B','C'] as `oldPath` and ['K','L'] as `addedPath` and creates `['A','B','C','']` 
  and `['A','B','C','K','L']` to pass them as second and third parameter for `nodeChange`

raineorshine replied:
In 2/3 cases (Line 260 and 272), `oldPath` and `addedPath` appear to be derivable from `newPath`. The third case (Line 267) is very close. I am looking for a way to have all three of them use the same interface.

What about this?

```js
const nodeAdd = (tree, newPath) => 
  nodeChange(
    tree, 
    newPath.concat({ value: DUMMY_TOKEN }), 
    newPath
  )
```

shresthabijay replied:
I understand your perspective but `addedPath` is not always derivable from `oldPath`. It's just in the case of line `260` and `272` that `oldPath` and `addedPath` are of same length. But in case of line `267`  the `addedPath` is longer path.

raineorshine replied:
Okay, just wanted to make sure. I think the arguments can still be derived from `commonPath`, but that would add a redundant call to `findTreeDeepestSubcontext` within `nodeMove`. I can live with this as-is for now. Thanks for engaging.

raineorshine commented on src/constants.js:
```diff
@@ -100,6 +100,9 @@ export const ROOT_TOKEN = '__ROOT__'
 export const EM_TOKEN = '__EM__'
 
+// token for recentlyEdited tree nodeChange special case (add)
+export const DUMMY_TOKEN = ''
```
It would be better to use a value of `__DUMMY__` as it is less likely to conflict with real values.

This can also be defined in `recentlyEditedTree.js` since it is only used in that one file.

shresthabijay replied:
Sure

raineorshine replied:
I meant `const DUMMY_TOKEN = '__DUMMY__'`. Sorry for the confusion.

shresthabijay replied:
Done!

raineorshine commented on src/util/sync.js:
```diff
@@ -64,7 +64,6 @@ export const sync = (thoughtIndexUpdates = {}, contextIndexUpdates = {}, { local
-
     // firebase
     if (remote) {
       return syncRemote(thoughtIndexUpdates, contextIndexUpdates, recentlyEdited, updates, callback)
```
I'm getting the following error when trying to sync to Firebase:

```js
Reference.update failed: First argument contains an invalid key (3/11) in property 
'users.m9S244ovF7fVrwpAoqoWxcz08s52.recentlyEdited.📒 Journal.2020:'. Keys must be non-empty strings 
and can't contain ".", "#", "$", "/", "[", or "]"
```

We need to `firebaseEncode` and `firebaseDecode` the keys in the tree in order to save it to Firebase. This isn't needed with `thoughtIndex` and `contextIndex` since the keys are already hashed.

Option 1: Stringify the entire `recentlyEdited` tree and save the whole tree to Firebase on each edit (inefficient)
Option 2: Encode they keys and only update the changed nodes in the tree on edit.

You can test this out by logging in via the Log In link in the footer and trying to save a thought with an invalid character such as "3/11".

shresthabijay replied:
Agreed! Option 1 will be very inefficient. Let's encode provided `oldContext` and `newContext` for all the `recentlyEditedTree` methods. All the nodes will have encoded keys . We won't even have to decode the keys of tree Object to render recently edited list because `path` property stored by leaf nodes won't be encoded.

shresthabijay commented:
In this latest commit I have modified `nodeAdd` to only use single parameter as we discussed earlier . I also implemented context encoding for escaping unsafe characters including empty string for firebase sync. Please review the changes!
